doc/foreign-function-interface.md

```diff
@@ -0,0 +1,341 @@
+
+```haskell
+foreign import javascript unsafe "$r1 = $1; $r2 = 0;"
+  js_mkPtr :: JSVal -> Ptr a
+
+foreign import javascript unsafe "$r = $1_1;"
+  js_ptrVal  :: Ptr a -> JSVal
+```
+
+### Caveats on Javascript FFI
+
+* `Int64` and `Word64` can be manipulated naturally in haskell code, and `Int64` and
+`Word64` can be passed into and returned from javascript functions via
+Javascript FFI. However, since javascript doesn't support `Int64` and `Word64`,
+they are represented as size-2 numerical types in javascript, and it'll be very very difficult to manipulate size-2 numerical types in javascript functions. Therefore, it's *strongly* recommended that you use `Double` as 53-bit integer instead of `Int64` and `Word64` in Javascript FFI.
+* It's ok to pass `Float` values into Javascript FFI, and `Float` can be returned
+from Javascript FFI. But, since javascript has only `Double`. Javascript FFI
+can return `Double` values exceeding the limitation of `Float`. Thus, it's
+recommended that you return `Double` or `JSVal` instead of `Float` in Javascript FFI.
+* You won't need to use `Ptr a`, `FunPtr a`, or `StablePtr a` in Javascript FFI in most cases because they exist to interface with C code.
+
+# Calling Haskell From Javascript
+
+It is possible to call haskell from javascript. But, as far as I know, there is no clean way to call haskell from javascript, yet. It's better to not think about it for now. If you think calling GHCJS from javascript is doable, feel free to modify this section.
+
+Right now, if you want to write javascript modules or javascript plugins in haskell, you should consider PureScript. It is similar to haskell and interfaces seamlessly with javascript.
+
+# Adding global javascript functions to Cabal Packages
+
+This section assumes that you have a basic understanding of cabal. GHCJS FFI has access to the global javascript variables. If you want to add more javascript functions to the global scope, you can add javascript files to `js-sources` in `.cabal` files as below.
+
+```
+library
+  js-sources:      jsbits/array.js
+                   jsbits/animationFrame.js
+                   jsbits/export.js
+                   jsbits/jsstring.js
+                   jsbits/jsstringRaw.js
+                   jsbits/foreign.js
+                   jsbits/text.js
+                   jsbits/utils.js
+                   jsbits/xhr.js
+                   jsbits/websocket.js
+```
+
+When you upload your GHCJS library or program to hackage, the javascript files in `js-sources` will be uploaded to hackage, too.
+It's like `c-sources` for including snippets of C code in cabal packages. Strangely, `CPP` extension is enabled by default in `js-sources`, so you can include a C header in `js-sources`.
```

Comment:
I think the explanation can be clearer, but I think your explanation is probably better than mine. Feel free to modify this before or after merging it.